scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -407,107 +417,232 @@
-            (not (compare-and-set! pending-jobs scheduler-contents new-scheduler-contents))
-            (do (log/info "Pending job atom contention encountered")
+      offer-size-mem
+      (get-in offer [:resources :mem] 0)))
+  (log/debug "invoked handle-resource-offers!")
+  (let [offer-stash (atom nil)] ;; This is a way to ensure we never lose offers fenzo assigned if an errors occures in the middle of processing
+    (timers/time!
+      handle-resource-offer!-duration
+      (try
+        (loop [] ;; This loop is for compare-and-set! below
+          (let [scheduler-contents @pending-jobs
+                db (db conn)
+                _ (log/debug "There are" (count scheduler-contents) "pending jobs")
+                considerable (->> scheduler-contents
+                                  (filter (fn [job]
+                                            (job-allowed-to-start? db job)))
+                                  (take num-considerable))
+                _ (log/debug "We'll consider scheduling" (count considerable) "of those pending jobs (limited to " num-considerable " due to backdown)")
+                matches (match-offer-to-schedule fenzo considerable offers db fid)
+                _ (reset! offer-stash (doall (for [{:keys [leases]} matches
+                                                   lease leases]
+                                               (:offer lease))))
+                matched-jobs (for [match matches
+                                   ^TaskAssignmentResult task-result (:tasks match)
+                                   :let [task-request (.getRequest task-result)]]
+                               (:job task-request))
+                processed-matches (process-matches-for-backfill scheduler-contents (first considerable) matched-jobs)
+                new-scheduler-contents (remove (fn [pending-job]
+                                                 (contains? (:fully-processed processed-matches) (:job/uuid pending-job)))
+                                               scheduler-contents)
+                first-considerable-resources (-> considerable first util/job-ent->resources)
+                match-resource-requirements (util/sum-resources-of-jobs matched-jobs)]
+            (reset! front-of-job-queue-mem-atom
+                    (or (:mem first-considerable-resources) 0))
+            (reset! front-of-job-queue-cpus-atom
+                    (or (:cpus first-considerable-resources) 0))
+            (cond
+              (empty? matches) (if (seq considerable)
+                                 nil ;; This case means we failed to match anything. We need to start head of line blocking now
+                                 true) ;; This case means there was nothing to match--don't penalize the scheduler for the next burst of submitted jobs
+              (not (compare-and-set! pending-jobs scheduler-contents new-scheduler-contents))
+              (do
```

Comment:
If I remember correctly, this code deals with multiple thread running handle-resource-offers! How does multiple thread  with a single fenzo instance work? Do we still to need to handle multiple thread case or it will be single threaded?